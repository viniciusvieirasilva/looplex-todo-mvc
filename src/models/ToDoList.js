import { types, getParent, destroy } from 'mobx-state-tree'
import { Howl } from 'howler'

const sound = new Howl({
  src: ['../../assets/comp.wav']
})

export const ToDoListItem = types
  .model({
    id: types.string,
    title: types.string,
    isCompleted: false
  })
  .actions(self => ({
    changeTitle (newTitle) {
      if (newTitle != '') {
        self.title = newTitle
      }
    },
    toggleIsCompleted () {
      self.isCompleted = !self.isCompleted
      //   if (self.isCompleted && !sound.playing()) {
      if (self.isCompleted) {
        if (sound.playing()) {
          sound.stop()
          sound.play()
        } else {
          sound.play()
        }
      }
      //   }
    },
    removeItem () {
      getParent(self, 2).remove(self)
    }
  }))

export const ToDoList = types
  .model({
    items: types.optional(types.array(ToDoListItem), [])
  })
  .actions(self => ({
    add (title) {
      if (title != '') {
        self.items.push({
          id: encodeURIComponent(title.toLowerCase().replace(/\s+/gim, '-')),
          title: title,
          isCompleted: false
        })
      }
    },
    remove (item) {
      destroy(item)
    },
    toggleListIsCompleted () {
      if (self.items.some(item => !item.isCompleted)) {
        self.items.forEach(item => {
          if (!item.isCompleted) {
            item.toggleIsCompleted()
          }
        })
      } else {
        self.items.forEach(item => item.toggleIsCompleted())
      }
    },
    removeCompleted () {
      if (self.items.some(item => item.isCompleted)) {
        self.items.forEach(item => {
          if (item.isCompleted) {
            item.removeItem()
          }
        })
      }
    }
  }))
  .views(self => ({
    get totalCount () {
      return self.items.length
    },
    get totalLeftCount () {
      return self.items.filter(item => !item.isCompleted).length
    },
    get totalCompletedCount () {
      return self.items.filter(item => item.isCompleted).length
    },
    getfilteredItems (filter) {
      const states = {
        todos: self.items,
        concluidos: self.items.filter(item => item.isCompleted),
        naoConcluidos: self.items.filter(item => !item.isCompleted)
      }
      return states[filter]
    }
  }))
