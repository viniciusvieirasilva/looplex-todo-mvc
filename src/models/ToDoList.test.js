import { ToDoList, ToDoListItem } from './ToDoList'
import { getSnapshot, onPatch } from 'mobx-state-tree'

it('can create a instance of a model', () => {
  const item = ToDoListItem.create({
    id: 'item 1',
    title: 'Tarefa 1',
    isCompleted: false
  })

  expect(item.title).toBe('Tarefa 1')
  item.changeTitle('Tarefa 3')
  expect(item.title).toBe('Tarefa 3')
  expect(item.isCompleted).toBe(false)
  item.toggleIsCompleted()
  expect(item.isCompleted).toBe(true)
})

it('can create a ToDoList, add and remove items', () => {
  const list = ToDoList.create({
    items: [
      {
        id: 'item-1',
        title: 'item 1',
        isCompleted: false
      },
      {
        id: 'item-2',
        title: 'item 2',
        isCompleted: true
      }
    ]
  })

  const patches = []
  onPatch(list, snapshot => {
    patches.push(snapshot)
  })

  expect(list.items.length).toBe(2)
  expect(list.items[1].title).toBe('item 2')

  list.items[1].removeItem()
  expect(list.items.length).toBe(1)

  list.add('item 3')
  expect(list.items.length).toBe(2)

  expect(getSnapshot(list)).toMatchSnapshot()

  expect(patches).toMatchSnapshot()
})
