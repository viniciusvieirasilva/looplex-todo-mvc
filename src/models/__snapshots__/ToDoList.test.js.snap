// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a ToDoList, add and remove items 1`] = `
Object {
  "items": Array [
    Object {
      "id": "item-1",
      "isCompleted": false,
      "title": "item 1",
    },
    Object {
      "id": "item-3",
      "isCompleted": false,
      "title": "item 3",
    },
  ],
}
`;

exports[`can create a ToDoList, add and remove items 2`] = `
Array [
  Object {
    "op": "remove",
    "path": "/items/1",
  },
  Object {
    "op": "add",
    "path": "/items/1",
    "value": Object {
      "id": "item-3",
      "isCompleted": false,
      "title": "item 3",
    },
  },
]
`;
